### PR TITLE
Adding a policy engine simulator

### PR DESCRIPTION
**Pull request by matajoh**
Draft PR.

### Changes
- Interaction with Rego has been moved out into its own extractable package: `regopolicyinterpreter`
- `regopolicyinterpreter` exposes various Rego things like modules and metadata in a typed way to make them easier to work with:
    - `RegoPolicyInterpreter` is the main interface
    - `RegoModule` is a standalone Rego module that can be included in the policy execution. 
     There are `AddModule` and `RemoveModule` for modifying the interpreter to include various modules.
    - `RegoQueryResult` wraps the results that come from the Rego policy with some useful methods for extracting
      scalar data types (bool/int/float/string)
- `policyenginesimulator` tool which uses `RegoPolicyInterpreter` to simulate policy enforcement.

### CODE REVIEW

File: pkg/regopolicyinterpreter/framework.rego [hunk 1]
```diff
@@ -177,14 +177,14 @@ create_container := {"matches": matches, "started": started, "allowed": true} {
 }
 
 mountSource_ok(constraint, source) {
-    startswith(constraint, data.sandboxPrefix)
```
Comment (SeanTAllen) [inline pkg/regopolicyinterpreter/framework.rego:180]: why the change in sandboxPrefix location in the `data` object?

consider this a question for all the other changes like this as well hugepages, default mounts etc.

File: pkg/securitypolicy/regopolicy_test.go [hunk 2]
```diff
@@ -1050,8 +1050,8 @@ func Test_Rego_Enforcement_Point_Allowed(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	input := make(map[string]interface{})
-	allowed, err := policy.allowed("__fixture_for_allowed_test_false__", input)
+	result := &rpi.RegoQueryResult{IsEmpty: true}
```
Comment (SeanTAllen) [inline pkg/securitypolicy/regopolicy_test.go:1053]: why the change from input to result? this feels meaningful in a way that I'm not grasping.

File: pkg/securitypolicy/regopolicy_test.go [hunk 3]
```diff
@@ -1023,17 +1023,17 @@ func Test_Rego_Version_Unavailable_Enforcement_Point(t *testing.T) {
...
 
-	info, err := policy.queryEnforcementPoint("__fixture_for_allowed_test_true__")
+	info, err := policy.rego.GetEnforcementPointInfo("__fixture_for_allowed_test_true__")
 	// we do not expect an error, getting one means something is broken
 	if err != nil {
 		t.Fatalf("unable to query whether enforcement point is available: %v", err)
 	}
 
-	if info.availableByPolicyVersion {
+	if info.AvailableByPolicyVersion {
```
Comment (SeanTAllen) [inline pkg/securitypolicy/regopolicy_test.go:1032]: this change seems to say "this is a engine test and will go into the engine's tests not as an Enforcement test", do you agree?

File: pkg/regopolicyinterpreter/framework.rego (hunk 1 above)
Comment (matajoh) [reply pkg/regopolicyinterpreter/framework.rego:180]: There are some clear benefits:
- All the interpreter needs to model is the metadata structure.
- Frees up the base `data` namespace such that the only part of that namespace we need is `metadata`. 
 
On the other hand, it means a customer policy could modify those values using metadata commands, which is probably a mixed blessing. Anyway, it seemed like the cleanest interface boundary.

File: pkg/securitypolicy/regopolicy_test.go (hunk 2 above)
Comment (matajoh) [reply pkg/securitypolicy/regopolicy_test.go:1053]: This is the latest episode of fun with types. The original version of `allowed` took the input. At some point we changed it to take the result set, but as neither was typed, it assumed the empty input of the original was an empty result set and thus no compile error and everything kept working. Once the query result became typed, this popped up as an error and I fixed it.

File: pkg/securitypolicy/regopolicy_test.go (hunk 3 above)
Comment (matajoh) [reply pkg/securitypolicy/regopolicy_test.go:1032]: Yup, these tests almost certainly should move to the new module. Before I start moving tests though I want to have the conversation about whether the current extraction is correct, as that will determine which tests should move.

File: pkg/regopolicyinterpreter/framework.rego [hunk 4]
```diff
@@ -177,14 +177,14 @@ create_container := {"matches": matches, "started": started, "allowed": true} {
...
-    newConstraint := replace(constraint, data.sandboxPrefix, input.sandboxDir)
+    startswith(constraint, data.metadata._mountinfo_.sandboxPrefix)
+    newConstraint := replace(constraint, data.metadata._mountinfo_.sandboxPrefix, input.sandboxDir)
```
Comment (SeanTAllen) [inline pkg/regopolicyinterpreter/framework.rego:181]: why has everything moved under "data"? I find the old way much easier to understand.

Its now metadata because?

And why the odd python-ism `_mountinfo_` with the `_` on either side?

File: internal/tools/policyenginesimulator/main.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	err = json.Unmarshal(contents, &ops)
+	if err != nil {
+		log.Fatalf("error unmarshalling metadata operations: %v", err)
+	}
+
+	err = r.ApplyAll(ops)
+	if err != nil {
+		log.Fatalf("error applying setup metadata operations: %v", err)
+	}
+}
```
Comment (SeanTAllen) [inline internal/tools/policyenginesimulator/main.go:125]: So in order to do the testing we want, we have to expose the metadata and then set it up on each side right? That feels very wrong. It metadata is purely an internal concern. Or rather it should be. 

This statement goes against what we discussed the other day, but it feels wrong now. I think we can do better.

Comment (matajoh) [reply internal/tools/policyenginesimulator/main.go]: The abstraction boundary has moved significantly since this comment (as discussed) so closing the convo.

File: pkg/regopolicyinterpreter/framework.rego (hunk 4 above)
Comment (matajoh) [reply pkg/regopolicyinterpreter/framework.rego]: Back the way it was, disregard.

File: pkg/securitypolicy/regopolicy_test.go (hunk 3 above)
Comment (matajoh) [reply pkg/securitypolicy/regopolicy_test.go]: With the abstraction boundary where it is now, it actually makes sense for these tests to stay put (i.e. the simulator is now very general purpose, and doesn't know anything about APIs or frameworks.)

File: pkg/regopolicyinterpreter/framework.rego (hunk 1 above)
Comment (matajoh) [reply pkg/regopolicyinterpreter/framework.rego]: back the way it was.
